### PR TITLE
[Buttons] Fix fallback behavior for Dynamic Type.

### PR DESCRIPTION
**Pull request by romoore**
The `titleFontForState:` API did not correctly reflect the rendered font
when no explicit fonts were set via the MDCButton APIs. As a result,
calling `updateTitleFont` internally could result in a different font
being returned than what was displayed.

Moving the "fall back" and Dynamic Type behavior into the
`titleFontForState:` method allows more consistent behavior (reduces the
risk of deviations) and corrects the APIs to reflect what is rendered.
This change should result in no visible changes to clients (the rendered
fonts should remain the same), but will result in non-`nil` results
being returned from `titleFontForState:` to correctly reflect the
rendered font for a given state.

Part of #7492

### CODE REVIEW

File: components/Buttons/tests/unit/ButtonsTests.m [hunk 1]
```diff
@@ -521,6 +521,70 @@ - (void)testTitleFontForStateFallbackBehavior {
   }
 }
 
+- (void)testTitleFontForStateFallbackBehaviorWithLegacyDynamicType {
+  // Given
```
Comment (codeman7) [inline components/Buttons/tests/unit/ButtonsTests.m:525]: _Tip:_ Fix this

**Review by codeman7 — CHANGES_REQUESTED** [review]

**Review by bryanoltman — APPROVED** [review]

File: components/Buttons/tests/unit/ButtonsTests.m (hunk 1 above)
Comment (romoore) [reply components/Buttons/tests/unit/ButtonsTests.m:525]: Done.

File: components/ButtonBar/tests/unit/ButtonBarButtonTitleFontTests.swift [hunk 2]
```diff
@@ -32,7 +33,7 @@ class ButtonBarButtonTitleFontTests: XCTestCase {
     // Then
     for view in buttonBar.subviews {
       if let button = view as? MDCButton {
-        XCTAssertNil(button.titleFont(for: .normal))
+        XCTAssertTrue(button.titleFont(for: .normal)!.mdc_isSimplyEqual(MDCTypography.buttonFont()))
```
Comment (codeman7) [inline components/ButtonBar/tests/unit/ButtonBarButtonTitleFontTests.swift]: _Tip:_ Safely unwrap this in an `if let` and then in the `else` print out a logical error message.

File: components/NavigationBar/tests/unit/NavigationBarButtonTitleFontTests.swift [hunk 3]
```diff
@@ -52,7 +53,7 @@ class NavigationBarButtonTitleFontTests: XCTestCase {
     // Then
     for view in recursiveSubviews(of: navigationBar) {
       if let button = view as? MDCButton {
-        XCTAssertNil(button.titleFont(for: .normal))
+        XCTAssertTrue(button.titleFont(for: .normal)!.mdc_isSimplyEqual(MDCTypography.buttonFont()))
```
Comment (codeman7) [inline components/NavigationBar/tests/unit/NavigationBarButtonTitleFontTests.swift]: _Tip:_ Same as above.

**Review by codeman7 — APPROVED** [review]

**Comment by codeman7** [comment]
@romoore Can you add a test where `adjust...Unavailabe = NO` and `mdc_adjust... = YES`.

**Comment by romoore** [comment]
@codeman7 Isn't that `testLegacyDynamicTypeDisabledThenDynamicTypeTurnedOn`? Is there something in that test you think it should verify further?

File: components/NavigationBar/tests/unit/NavigationBarButtonTitleFontTests.swift (hunk 3 above)
Comment (romoore) [reply components/NavigationBar/tests/unit/NavigationBarButtonTitleFontTests.swift]: Done

File: components/ButtonBar/tests/unit/ButtonBarButtonTitleFontTests.swift (hunk 2 above)
Comment (romoore) [reply components/ButtonBar/tests/unit/ButtonBarButtonTitleFontTests.swift]: Done

**Comment by codeman7** [comment]
@romoore Whoops I just saw the asserts, I didn't expand to see the Given, When. LGTM!

**Pull request merged**